src/libaktualizr/primary/sotauptaneclient.cc

```diff
@@ -552,38 +552,76 @@ bool SotaUptaneClient::updateImagesMeta() {
   // Update Images Targets Metadata
   {
     std::string images_targets;
+    Uptane::Role targets_role = Uptane::Role::Targets();
 
     int64_t targets_size = (images_repo.targetsSize() > 0) ? images_repo.targetsSize() : Uptane::kMaxImagesTargetsSize;
     if (!uptane_fetcher->fetchLatestRole(&images_targets, targets_size, Uptane::RepositoryType::Image(),
-                                         Uptane::Role::Targets())) {
+                                         targets_role)) {
       return false;
     }
     int remote_version = Uptane::extractVersionUntrusted(images_targets);
 
     int local_version;
     std::string images_targets_stored;
-    if (storage->loadNonRoot(&images_targets_stored, Uptane::RepositoryType::Image(), Uptane::Role::Targets())) {
+    if (storage->loadNonRoot(&images_targets_stored, Uptane::RepositoryType::Image(), targets_role)) {
       local_version = Uptane::extractVersionUntrusted(images_targets_stored);
     } else {
       local_version = -1;
     }
 
-    if (!images_repo.verifyTargets(images_targets)) {
+    if (!images_repo.verifyTargets(images_targets, targets_role)) {
       last_exception = images_repo.getLastException();
       return false;
     }
 
     if (local_version > remote_version) {
       return false;
     } else if (local_version < remote_version) {
-      storage->storeNonRoot(images_targets, Uptane::RepositoryType::Image(), Uptane::Role::Targets());
+      storage->storeNonRoot(images_targets, Uptane::RepositoryType::Image(), targets_role);
     }
 
-    if (images_repo.targetsExpired()) {
+    if (images_repo.targetsExpired("targets")) {
       last_exception = Uptane::ExpiredMetadata("repo", "targets");
       return false;
     }
   }
+
+  // Update Images delegated Targets Metadata
+  for (const std::string &delegation : images_repo.delegations("targets")) {
```

Comment:
IIRC @heartsucker had an idea that the delegation tree can be huge and we only want to download what is relevant on this particular update. Also one of the drivers to add support for delegations now was the idea to save space/bandwidth on potentially huge `targets.json` for maps by splitting them by region.

So something like an optional (to also support tuf-only scenario) parameter specifying list of targets from director would be useful. Or as an alternative (and probably closer to the spirit of TUF spec) solution, we could only download delegations when intending to download a target or are explicitly asked to.

It may better be left for a next PR though.

Reply:
I recall the same conversations and had similar concerns. For the moment, we've "avoided" the problem by only processing first-order delegations. On one hand, I'm not sure how likely it really is that we will have more than a couple layers of delegation in the automotive world, but on the other hand I think it makes sense to be flexible with our design goals and thus careful and efficient in our implementation.

I agree that we probably shouldn't download all delegations and instead only get what we need. However, that would require a fairly substantial refactoring and I wasn't willing to do it for this PR. Perhaps that should be a requirement for the nested delegation support, though.

Reply:
@patrickvacek 
I think these issues somewhat correlate, but we can also run into bandwidth problems with simple non-nested delegations. For the aforementioned example of region-based delegation for maps, these don't necessarily need to be nested.

Reply:
Fair point. Sounds like a higher-priority chunk of work than nesting, then. It shouldn't be too hard, but since I'm unlikely to be able to get it done in the next week or so, would it be possible to merge this as it is and then do that as a follow-up? Others are welcome to contribute; I'm just a bit loathe to let this languish for a week while I'm away.

In other news, it sounds like the backend puts delegated_targets.json into a special `delegations/` subdirectory in the images repo.